File: backend/modules/petclinic/app/petclinic/api/pettype/getPetType/GetPetTypeService.java
package petclinic.api.pettype.getPetType;

import akka.actor.ActorRef;
import petclinic.api.Constants;
import petclinic.api.model.PetType;
import petclinic.model.pettypes.PetTypeProtocol;

import javax.inject.Inject;
import javax.inject.Named;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

public class GetPetTypeService {


    @Inject
    @Named("pettypes")
    private ActorRef petTypesActor;


    public CompletionStage<Optional<PetType>> getPetType(String id) {
        final PetTypeProtocol.GetQuery message = PetTypeProtocol.get(id);
        return akka.pattern.PatternsCS.ask(petTypesActor, message, Constants.TIMEOUT)
                .thenApply(obj -> (PetTypeProtocol.GetQueryResponse) obj)
                .thenApply(resp -> {
                    if (resp instanceof PetTypeProtocol.NotFoundGetQueryResponse) {
                        return Optional.empty();
                    } else if (resp instanceof PetTypeProtocol.FoundGetQueryResponse) {
                        final PetTypeProtocol.FoundGetQueryResponse resp1 = (PetTypeProtocol.FoundGetQueryResponse) resp;
                        return Optional.of(new PetType(resp1.id(), resp1.name()));
                    } else {
                        throw new IllegalStateException("unkown response type");
                    }
                });
    }

}

File: backend/modules/petclinic/app/petclinic/api/pettype/getPetType/GetPetTypeService.scala
package petclinic.api.pettype.getPetType

import akka.actor.ActorRef
import javax.inject.{Inject, Named}
import petclinic.api.Constants
import petclinic.api.model.PetType
import petclinic.model.pettypes.PetTypeProtocol
import petclinic.model.pettypes.PetTypeProtocol.{FoundGetQueryResponse, GetQueryResponse, NotFoundGetQueryResponse}

import scala.concurrent.{ExecutionContext, Future}

class GetPetTypeService @Inject()(@Named(petclinic.api.Constants.PET_TYPES_ACTOR_NAME) actor: ActorRef)(implicit ec: ExecutionContext) {

  def getPetType(idIn: String): Future[Option[PetType]] = {
    akka.pattern.Patterns.ask(actor, PetTypeProtocol.GetQuery(idIn), Constants.TIMEOUT)
      .mapTo[GetQueryResponse]
      .map {
        case FoundGetQueryResponse(id, name) => Some(new PetType(id, name))
        case NotFoundGetQueryResponse(id) => None
      }
  }

}
